### DISCUSSION TITLE
Run go-xdp-counter with Error "Database entry num_extension does not exist in tree xdp_dispatcher__2_1"

### DISCUSSION THREAD

**Discussion by ivzhh** [bpfman/bpfman#1206] [General]
Hi community, 

I have some issue on running the tutorial of bpfman. I am using a VM, running Ubuntu `Linux ws 6.8.0-39-generic #39-Ubuntu SMP PREEMPT_DYNAMIC Fri Jul  5 21:49:14 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux`. On either branch `v0.4.2` or `main`, I get the error of `"Database entry num_extension does not exist in tree xdp_dispatcher__2_1"`.

Run bpfman-rpc in one terminal:
```
sudo RUST_BACKTRACE=1 RUST_LOG=info ./target/debug/bpfman-rpc --timeout=0

[INFO  bpfman_rpc::serve] Using no inactivity timer
[INFO  bpfman_rpc::serve] Using default Unix socket
[INFO  bpfman_rpc::serve] Listening on /run/bpfman-sock/bpfman.sock
[INFO  bpfman::utils] Has CAP_BPF: true
[INFO  bpfman::utils] Has CAP_SYS_ADMIN: true
[INFO  bpfman::oci_utils::cosign] Starting Cosign Verifier, downloading data from Sigstore TUF repository
[INFO  bpfman::types] Loading program bytecode from file: /home/mao/repos/bpfman/examples/go-xdp-counter/bpf_x86_bpfel.o
thread 'tokio-runtime-worker' panicked at bpfman/src/multiprog/mod.rs:111:18:
failed to get xdp_dispatcher num_extensions: DatabaseError("Database entry num_extension does not exist in tree \"xdp_dispatcher__2_1\"", "")
```

Load bpf program in another
```
sudo ./go-xdp-counter -iface enp1s0

2024/07/29 17:50:13 Using Input: Interface=enp1s0 Priority=50 Source=/home/mao/repos/bpfman/examples/go-xdp-counter/bpf_x86_bpfel.o
2024/07/29 17:50:16 rpc error: code = Canceled desc = stream terminated by RST_STREAM with error code: CANCEL
```

My guess is the default kernel in Ubuntu 24.04 doesn't support `xdp_dispatcher__2_1` extension. I don't find any guidelines for setting up the right kernel to run bpfman. May I get some advices here? 

Thank you!

**Comment by Billy99** [comment]
Sorry for the delay, we had several maintainers out last week.

I just ran the latest code on an Ubuntu 23.10 VM with 6.5.0-28-generic and it worked fine. So I don't think it's your kernel version. `xdp_dispatcher__2_1` is a database entry name, where in your case ,the `_2` is the ifindex and the `_1` is the revision. The error `Database entry num_extension does not exist in tree \"xdp_dispatcher__2_1\"` indicates the field `num_extension` was not in the database.

I'm not sure what went wrong, I'll play with it and see if I can reproduce. In the mean time, there are two things you can try:
1) Restart from scratch by removing all installed files and restarting everything. To do that, CTRL-C out of the running bpfman-rpc command if it is still running, then run `sudo ./scripts/setup.sh uninstall` to cleanup. Then retry.

2) Trying running as a systemd service. That's where most of the testing is occurring. Instead of running  `sudo RUST_BACKTRACE=1 RUST_LOG=info ./target/debug/bpfman-rpc --timeout=0`, run:
`sudo ./scripts/setup.sh uninstall`
`sudo ./scripts/setup.sh install`
Then use `sudo journalctl -f -u bpfman` to see the logs.

If you know of any steps you did that might help me reproduce the problem, let me know.

**Reply by ivzhh** [reply]
Thanks Billy for the help, after using systemd service for testing, I can still see the same issue. The Ubuntu is a guest VM atop an Alma Linux, using bridged network. Therefore, the Ubuntu can get the router IP.

`Linux ws 6.8.0-39-generic #39-Ubuntu SMP PREEMPT_DYNAMIC Fri Jul  5 21:49:14 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux` 

I am using bpfman `tag: v0.4.2 eef28663a59f4607675f79e2e2ed6e78148969c4` with one change: using multiple `-I` to include header paths (only this works on my Ubuntu).

```
diff --git a/examples/go-xdp-counter/main.go b/examples/go-xdp-counter/main.go
index 812f22a9..eca79187 100644
--- a/examples/go-xdp-counter/main.go
+++ b/examples/go-xdp-counter/main.go
@@ -36,7 +36,7 @@ const (
        XDP_ACT_OK = 2
 )
 
-//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -cc clang -no-strip -cflags "-O2 -g -Wall" bpf ./bpf/xdp_counter.c -- -I.:/usr/include/bpf:/usr/include/linux
+//go:generate go run github.com/cilium/ebpf/cmd/bpf2go -cc clang -no-strip -cflags "-O2 -g -Wall" -target amd64,arm64,ppc64le,s390x bpf ./bpf/xdp_counter.c -- -I.:/usr/include/bpf -I/usr/include/linux -I/usr/include/x86_64-linux-gnu/
 func main() {
        stop := make(chan os.Signal, 1)
        signal.Notify(stop, os.Interrupt, syscall.SIGINT, syscall.SIGTERM)
```

Here is the testing output using systemd job:

```
sudo journalctl -f -u bpfman                          
Aug 05 21:48:24 ws bpfman-rpc[30941]: Has CAP_BPF: true
Aug 05 21:48:24 ws bpfman-rpc[30941]: Has CAP_SYS_ADMIN: true
Aug 05 21:48:24 ws bpfman-rpc[30941]: Starting Cosign Verifier, downloading data from Sigstore TUF repository
Aug 05 21:48:26 ws bpfman-rpc[30941]: Loading program bytecode from file: /home/mao/repos/bpfman/examples/go-xdp-counter/bpf_x86_bpfel.o
Aug 05 21:48:26 ws bpfman-rpc[30941]: thread 'tokio-runtime-worker' panicked at bpfman/src/multiprog/mod.rs:111:18:
Aug 05 21:48:26 ws bpfman-rpc[30941]: failed to get xdp_dispatcher num_extensions: DatabaseError("Database entry num_extension does not exist in tree \"xdp_dispatcher__2_1\"", "")
Aug 05 21:48:26 ws bpfman-rpc[30941]: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Aug 05 21:48:39 ws bpfman-rpc[30941]: Shutdown Unix Handler /run/bpfman-sock/bpfman.sock
Aug 05 21:48:39 ws systemd[1]: bpfman.service: Deactivated successfully.
Aug 05 21:48:39 ws systemd[1]: bpfman.service: Consumed 1.091s CPU time.
Aug 05 21:49:27 ws systemd[1]: Started bpfman.service - Run bpfman as a service.
Aug 05 21:49:27 ws bpfman-rpc[31521]: Using a Unix socket from systemd
Aug 05 21:49:27 ws bpfman-rpc[31521]: Listening on /run/bpfman-sock/bpfman.sock
Aug 05 21:49:27 ws bpfman-rpc[31521]: Using inactivity timer of 15 seconds
Aug 05 21:49:27 ws bpfman-rpc[31521]: Has CAP_BPF: true
Aug 05 21:49:27 ws bpfman-rpc[31521]: Has CAP_SYS_ADMIN: true
Aug 05 21:49:27 ws bpfman-rpc[31521]: Starting Cosign Verifier, downloading data from Sigstore TUF repository
Aug 05 21:49:29 ws bpfman-rpc[31521]: Loading program bytecode from file: /home/mao/repos/bpfman/examples/go-xdp-counter/bpf_x86_bpfel.o
Aug 05 21:49:32 ws bpfman-rpc[31521]: The bytecode image: quay.io/bpfman/xdp-dispatcher:v2 is signed
Aug 05 21:49:42 ws bpfman-rpc[31521]: Shutdown Unix Handler /run/bpfman-sock/bpfman.sock
Aug 05 21:49:42 ws systemd[1]: bpfman.service: Deactivated successfully.
Aug 05 21:49:42 ws systemd[1]: bpfman.service: Consumed 1.790s CPU time.
Aug 05 21:49:50 ws systemd[1]: Started bpfman.service - Run bpfman as a service.
Aug 05 21:49:50 ws bpfman-rpc[31704]: Using a Unix socket from systemd
Aug 05 21:49:50 ws bpfman-rpc[31704]: Using inactivity timer of 15 seconds
Aug 05 21:49:50 ws bpfman-rpc[31704]: Listening on /run/bpfman-sock/bpfman.sock
Aug 05 21:49:50 ws bpfman-rpc[31704]: Has CAP_BPF: true
Aug 05 21:49:50 ws bpfman-rpc[31704]: Has CAP_SYS_ADMIN: true
Aug 05 21:49:51 ws bpfman-rpc[31704]: Starting Cosign Verifier, downloading data from Sigstore TUF repository
Aug 05 21:49:53 ws bpfman-rpc[31704]: Loading program bytecode from file: /home/mao/repos/bpfman/examples/go-xdp-counter/bpf_x86_bpfel.o
Aug 05 21:49:53 ws bpfman-rpc[31704]: thread 'tokio-runtime-worker' panicked at bpfman/src/multiprog/mod.rs:111:18:
Aug 05 21:49:53 ws bpfman-rpc[31704]: failed to get xdp_dispatcher num_extensions: DatabaseError("Database entry num_extension does not exist in tree \"xdp_dispatcher__2_1\"", "")
Aug 05 21:49:53 ws bpfman-rpc[31704]: note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Aug 05 21:50:05 ws bpfman-rpc[31704]: Shutdown Unix Handler /run/bpfman-sock/bpfman.sock
Aug 05 21:50:05 ws systemd[1]: bpfman.service: Deactivated successfully.
Aug 05 21:50:05 ws systemd[1]: bpfman.service: Consumed 1.242s CPU time.
```

```
sudo ./go-xdp-counter -iface enp1s0 -file $PWD/bpf_x86_bpfel.o
2024/08/05 21:49:50 Using Input: Interface=enp1s0 Priority=50 Source=/home/mao/repos/bpfman/examples/go-xdp-counter/bpf_x86_bpfel.o
2024/08/05 21:49:54 rpc error: code = Canceled desc = stream terminated by RST_STREAM with error code: CANCEL
```

**Comment by Billy99** [comment]
I loaded v0.4.2 and v0.5.0 see the same error are you are seeing, but not seeing on main. Trying to track down what might have been the original issue and what fixed it.

**Comment by Billy99** [comment]
I found the issue. Even though we have version tags for bpfman, the code is always loading the latest xdp-dispatcher image. So the bpfman v0.4.2 and bpfman v0.5.0 code is trying to load `quay.io/bpfman/xdp-dispatcher:v2` (which is essentially latest) and not able to load it because post v0.5.0 we removed/changed some labels associated with the bytecode image and the old code can't find the `io.ebpf.program_name` label.

Bottom line, older releases are currently broken for XDP and TC. I'm working on a fix and we will cut a v0.5.1 image as soon as we can. `main` is working if you are able to build it.

**Comment by Billy99** [comment]
I was able to upload an older xdp-dispatcher and tc-dispatcher to quay.io. So v0.4.2 and v0.5.0 should work now. Let me know if you still have issues.
